### ISSUE TITLE
GUI scaling (hiDPI displays on Linux)

### ISSUE THREAD

**Issue by tferr** [trakem2/TrakEM2#30]
@axtimwalde, @acardona : Just so that it is not forgotten: When doing this [test](https://github.com/scijava/pom-scijava/issues/199#issuecomment-1232063709), i noticed the following with trakEM2 (PopOS 22.04, default Java8 bundled with Fiji):
- The program does not respect IJ1's scaling factor for its toolbar (IJ1: Edit -> Options -> Appearance: GUI scale)
- It imposes the un-scalable metal L&F, which is unfortunate because it does not allow to use FlatLaf that does scale nicely with Java8 on 4K displays

The only way to use trakEM2 (on linux) in a somewhat comfortable manner is to rescale the screen to 1080p, which is really sad

**Comment by acardona** [comment]
Thanks @tferr, I am using TrakEM2 in Ubuntu and somehow it looks OK? My desktop screen is 2500x1400 or something like that.

The toolbar is broken, as ImageJ has changed something and TrakEM2 hasn't kept up, but that's another matter.

**Comment by axtimwalde** [comment]
That is medium resolution and workable.  However, we have 3456x2160px
on a 13" screen.  This is when it gets hard to see anything.  FlatLaf
has very nice scalable LoFs that deal with that but it can be hard if
there are too many pixel based design decisions in the way.  We just
did this for BigDataViewer and it was a bit of work.

**Comment by tferr** [comment]
>  However, we have 3456x2160px on a 13" screen. 

Similar to my laptop. It is just not trakEM2. Lots of things in Fiji (with Java 8) are affected by this. It just makes it hard to use it out of the box. I'm going to lobby for wider FlatLaf adoption. At least having it available in Fiji would allow scripts to make Swing-based GUIs (more) scalable even after being displayed.

**Comment by rasband** [comment]
> The toolbar is broken, as ImageJ has changed something and TrakEM2 hasn't kept up, but that's another matter.

The tool buttons are 2 pixels larger in ImageJ 1.53c and later so TrakEM2, when drawing its toolbar, needs to adjust the spacing to account for this change.

![Screenshot](https://user-images.githubusercontent.com/1633637/187591557-b57e1a5f-9070-4084-ba6e-452ebf86e7f3.png)

**Comment by tferr** [comment]
@acardona, @acardona 
Looked quickly at this (thanks for the tip @rasband). The toolbar was accessing a deprecaded `SIZE` field in `ij.gui.Toolbar`. Updating it to `buttonHeight` seems to fix the toolbar for the most part:
![Screenshot from 2022-08-31 09-03-55](https://user-images.githubusercontent.com/2439948/187690141-118e0445-8cca-474a-9353-bafdb650bfc6.png)

It is not perfect (notice how the second row is shifted to the left), but at least one can now interact with the tools :smile: 

I also tried to have them scale according to whatever scaling factor is in use in IJ's Edit->Options->Appearance: If users are scaling IJ's GUI (as I am on linux), than the tools will look the same across trakEM2 and IJ. I also made that the metal L&F is only imposed when Nimbus is being used. Overall, this allows to run TrakEM2 quite well on a 4K screen on Ubuntu. This is how it looks with FlatLaf:

![image](https://user-images.githubusercontent.com/2439948/187692496-eceb9e3b-f098-449f-9f22-fca764a07de2.png)

@rasband, related: `ij.Prefs.getGuiScale()` does not seem to be applied to the font of the contextual menus of the toolbar icons. They remain prohibitively small to read. Would it be possible to scale it too?

**Comment by acardona** [comment]
That's amazing! Thanks so much!

Missatge de Tiago Ferreira ***@***.***> del dia dc., 31 d’ag.
2022 a les 14:49:

**Comment by rasband** [comment]
Hi @tferr,
> It is not perfect (notice how the second row is shifted to the left), but at least one can now interact with the tools 

You probably need to use both `buttonWidth` and `buttonHeight` to position the tool icons.

> ij.Prefs.getGuiScale() does not seem to be applied to the font of the contextual menus of the toolbar icons.

I have not found a way to scale contextual menus, which are `popupMenu`s. I tried using setMenu() to increase the font size, but it did not work.

**Comment by tferr** [comment]
> I have not found a way to scale contextual menus

@rasband ,  [ij.gui.GUI#scalePopupMenu()](https://github.com/imagej/ImageJ/blob/d2f810ae85d2c0c17301a8567e57f08b09764e1a/ij/gui/GUI.java#L201https://github.com/imagej/ImageJ/blob/d2f810ae85d2c0c17301a8567e57f08b09764e1a/ij/gui/GUI.java#L201) should do this (I cannot remember: maybe I wrote it for the PlotWindow!?). It seems to work even when it is called right after initializing the PopupMenu, e.g., If one adds this to [ij.gui.Toolbar#addPopupMenus](https://github.com/imagej/ImageJ/blob/d2f810ae85d2c0c17301a8567e57f08b09764e1a/ij/gui/Toolbar.java#L166):

```java
		ovalPopup = new PopupMenu();
		ij.gui.GUI.scalePopupMenu(ovalPopup);
```

The font scales properly (compare to the unscaled font from the rectangular tool -- snapshot from the same ImageJ frame):
![image](https://user-images.githubusercontent.com/2439948/187778545-efa1e140-4929-4f4a-b5e9-afae7b8fc38b.png)

**Comment by rasband** [comment]
Hi @tferr, 

I had forgotten` ij.gui.GUI#scalePopupMenu()`! Unfortunately, it didn't appear to work when I added it to ij.gui.Toolbar#addPopupMenus() in the ImageJ 1.53u2 daily build. I tested it on macOS, Linux and Windows.

**Comment by rasband** [comment]
Hi @tferr,

Scaling of toolbar popup menus works on Linux and Windows in the ImageJ 1.53u4 daily build! The commit is at
https://github.com/imagej/ImageJ/commit/5fa55503bd668efc64900d2d80a8b7d22c840af7

**Comment by tferr** [comment]
Awesome @rasband, Glad it worked. But [Line 1795](https://github.com/imagej/ImageJ/blob/5fa55503bd668efc64900d2d80a8b7d22c840af7/ij/gui/Toolbar.java#L1795) still needs to be patched, otherwise all the menu tools from macros etc. will not scale. 

I will close this issue, as all this will make the trakEM2 GUI way more functional for a while.

**Comment by rasband** [comment]
Line 1795 of Toolbar.java is patched in the ImageJ 1.53u5 daily build. The commit is at https://github.com/imagej/ImageJ/commit/d7f396b7517fb990a6ad5d4f60d1efe3f27058c1.
